### DISCUSSION TITLE
Factory for creating rows in a table

### DISCUSSION THREAD

**Discussion by chrsoo** [Relm4/Relm4#689] [Q&A] [answered]
I would like to create a table based on a model that produces a row with multiple widgets, e.g. two labels, an entry, and a button. All widgets should be vertically aligned, producing a tabular format for the UI. Like a spreadsheet if you want, but I don't want visible cells.

Perhaps the solution is just to use a `gtk::Box` as the root and explicitly set the lengths of the widgets to the same in each column to make them align, but the problem seems to point to using a `gtk::Grid` as the parent widget of a factory.

If it is possible to use a grid as the parent, how do I position the individual widgets of each row given that the `FactroyComponent` trait and the `view!` macro produces a single root element?

Would it not mean that the view must produce more than one element instead of a single root widget?

Thanks!

**Comment by AaronErhardt** [comment]
I think you can use `gtk::Box`, actually. If you add the widgets to certain [size groups](https://gtk-rs.org/gtk4-rs/git/docs/gtk4/struct.SizeGroup.html), they will automatically have a uniform size across rows. To share the size groups more easily, you might use tricks like [`thread_local`](https://doc.rust-lang.org/std/macro.thread_local.html)

**Reply by chrsoo** [reply]
With this approach, I guess the parent widget of the factory is still a `gtk::Box` or similar, and not a `gtk::Grid`?

**Reply by chrsoo** [reply]
So... can't make `SizeGroup`s work with the `thread_local!` macro - running into issues with dereferencing the thread local, lifetime issues, and missing traits. For the last error (first below), I wonder if there might not be an issue with the view macro, not sure.

Any suggestions or help are welcome! 🙂

I have tried the following thread locals:
```rust
thread_local! {
    static PROP_ROW: PropRow = PropRow::default();
    static PROP_COLUMN_2: SizeGroup = SizeGroup::new(gtk::SizeGroupMode::Horizontal);
    static PROP_COLUMN_3: SizeGroup = SizeGroup::new(gtk::SizeGroupMode::Horizontal);
}
```
... where `PropRow` is an attempt to work around the issues with the more straight forward `SizeGroup` constants:

```rust
struct PropRow {
    column_1: SizeGroup,
    column_2: SizeGroup,
    column_3: SizeGroup,
}
impl Default for PropRow {
    fn default() -> Self {
        Self {
            column_1: SizeGroup::new(gtk::SizeGroupMode::Horizontal),
            column_2: SizeGroup::new(gtk::SizeGroupMode::Horizontal),
            column_3: SizeGroup::new(gtk::SizeGroupMode::Horizontal),
        }
    }
}
impl PropRow {
    fn column(&'static self, nbr: u8) -> &'static SizeGroup {
        match nbr {
            1 => &self.column_1,
            2 => &self.column_2,
            3 => &self.column_3,
            _ => unreachable!()
        }
    }
}
```

The factory `view!` declaration looks like the following:

```rust
    view! {
        #[root]
        gtk::Box {
            set_orientation: gtk::Orientation::Horizontal,
            set_spacing: 10,
            append = match &self.item.value {
                PropertyValue::Text(s) => {
                    gtk::Box {
                        gtk::Box {
                            set_size_group: PROP_ROW.column(1),
                            gtk::Label {
                                #[watch]
                                set_label: &self.item.name,
                            },
                        },
                        gtk::Box {
                            set_size_group: &PROP_COLUMN_2,
                            gtk::Label {
                                #[watch]
                                set_label: s,
                            },
                        },
                        gtk::Box {
                            set_size_group: PROP_COLUMN_3.with(|c| c),
                            gtk::Entry {
                                #[watch]
                                set_placeholder_text: Some(s),
                            }
                        },
                    }
                },
...
```
(Yes, I need the conditional widgets...)

(There are a lot of boxes around, but afaik it cannot be avoided as the `set_size_group` method is not available on labels, there must be a root element, and the conditional macro cannot be applied directly for the root element. Suggestions on how to reduce the number of boxes are welcome!)

The widget for the first colum that uses the `PropRow` struct
```rust
set_size_group: PROP_ROW.column(1)
```
... fails with the following compilation error:
```
error[E0599]: the method `column` exists for struct `LocalKey<PropRow>`, but its trait bounds were not satisfied
   --> relm/src/property.rs:143:54
    |
143 | ...                   set_size_group: PROP_ROW.column(1),
    |                                                ^^^^^^ method cannot be called on `LocalKey<PropRow>` due to unsatisfied trait bounds
    |
   ::: /Users/csoop/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:111:1
    |
111 | pub struct LocalKey<T: 'static> {
    | ------------------------------- doesn't satisfy `_: TreeViewExt` or `std::thread::LocalKey<PropRow>: IsA<TreeView>`
    |
    = note: the following trait bounds were not satisfied:
            `std::thread::LocalKey<PropRow>: IsA<TreeView>`
            which is required by `std::thread::LocalKey<PropRow>: gtk4::prelude::TreeViewExt`
    = help: use `with` or `try_with` to access thread local storage
note: the method `column` exists on the type `PropRow`
   --> relm/src/property.rs:24:5
    |
24  |     fn column(&'static self, nbr: u8) -> &'static SizeGroup {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
The widget for the second column using...
```rust
set_size_group: &PROP_COLUMN_2,
```
... fails with the following compilation error:
```
error[E0308]: mismatched types
   --> relm/src/property.rs:150:45
    |
150 | ...                   set_size_group: &PROP_COLUMN_2,
    |                       --------------  ^^^^^^^^^^^^^^ expected `&SizeGroup`, found `&LocalKey<SizeGroup>`
    |                       |
    |                       arguments to this method are incorrect
    |
    = note: expected reference `&SizeGroup`
               found reference `&std::thread::LocalKey<SizeGroup>`
note: method defined here
   --> /Users/csoop/.cargo/registry/src/index.crates.io-6f17d22bba15001f/relm4-0.9.1/src/extensions/widget_ext.rs:9:8
    |
9   |     fn set_size_group(&self, size_group: &gtk::SizeGroup);
```
The widget for the third column using...
```rust
set_size_group: PROP_COLUMN_3.with(|c| c)
```
... fails with the compilation error:
```
error: lifetime may not live long enough
   --> relm/src/property.rs:157:68
    |
157 | ...                   set_size_group: PROP_COLUMN_3.with(|c| c),
    |                                                           -- ^ returning this value requires that `'1` must outlive `'2`
    |                                                           ||
    |                                                           |return type of closure is &'2 SizeGroup
    |                                                           has type `&'1 SizeGroup`

```

**Reply by chrsoo** [reply]
Also tried sharing the size groups by adding them to the model, but here the issue is that I need them to be accessible globally, which I guess is the reason you suggested a thread local in the first place...

**Reply by AaronErhardt** [reply]
Sorry, my bad. I forgot that thread_local still has this ugly API what makes it very strange to work with (you can't just access the value easily due to some lifetime problems).
Anyway, a static `LazyLock` should solve that problem.

**Reply by chrsoo** [reply]
Thanks for the reply!

I am afraid that it does not resolve the problem though... Seems like `SizeGroup` uses raw pointers and that `LazyLock` requires the `Send` trait. The code
```rust
static PROP_COLUMN_1: LazyLock<SizeGroup> = LazyLock::new(|| {
    SizeGroup::new(gtk::SizeGroupMode::Horizontal)
});

```
... does not compile with the following error:
```
error[E0277]: `*mut c_void` cannot be sent between threads safely
  --> relm/src/property.rs:9:23
   |
9  | static PROP_COLUMN_1: LazyLock<SizeGroup> = LazyLock::new(|| {
   |                       ^^^^^^^^^^^^^^^^^^^ `*mut c_void` cannot be sent between threads safely
   |
   = help: the trait `Send` is not implemented for `*mut c_void`, which is required by `LazyLock<SizeGroup>: Sync`
   = note: required for `TypedObjectRef<*mut c_void, ()>` to implement `Sync`
note: required because it appears within the type `SizeGroup`
  --> /Users/csoop/.cargo/registry/src/index.crates.io-6f17d22bba15001f/gtk4-0.9.2/src/auto/size_group.rs:15:16
   |
15 |     pub struct SizeGroup(Object<ffi::GtkSizeGroup>) @implements Buildable;
   |                ^^^^^^^^^
   = note: required for `LazyLock<SizeGroup>` to implement `Sync`
   = note: shared static variables must have a type that implements `Sync`
```

**Reply by AaronErhardt** [reply]
Apparently, it only works with constants, not statics. I found an old comment of mine that I had in mind the whole time, but I messed up the details: https://github.com/gtk-rs/gtk-rs-core/pull/775

**Reply by chrsoo** [reply]
So... with the `LazyLocal`...
```rust
const PROP_COLUMN_1: LazyLock<SizeGroup> = LazyLock::new(|| {
    SizeGroup::new(gtk::SizeGroupMode::Horizontal)
});
```
... and the following assignment
```rust
set_size_group: &PROP_COLUMN_1,
```
... everything compiles. 

**However, the sizes of the widgets are all different. Using the `GTK_DEBUG=interactive` envar and checking in the GTK debug tool, they all seem to be different `SizeGroup` instances which would explain why the sizes are different.**

Reading up on `const` vs `static` I find the following snippet in the Rust documentation:
> `const` items looks remarkably similar to static items, which introduces some confusion as to which one should be used at which times. To put it simply, **constants are inlined wherever they're used**, making using them identical to simply replacing the name of the const with its value. Static variables, on the other hand, point to a single location in memory, which all accesses share. This means that, unlike with constants, they can't have destructors, and act as a single value across the entire codebase.

(added bold formatting to highlight the relevant text)

This makes me think that the `LazyLock` enclosure is called repeatedly as I guess that the LazyLock is "copy-pasted" by the compiler for the assignments and then executed only once when running the compiled binary. But perhaps I am getting this wrong.

Thoughts?

**Reply by AaronErhardt** [reply, marked as answer]
Mhh, we only progress from one problem to another. This is an particularly unpleasant problem in Rust and something I deal with rarely (which is why it took me so long to remember all the details).

I hope I've now considered all the potential problems. I think there are two possibilities:

+ Use `thread_local` and then `.with(|w| w.clone())` to access the widgets (since widgets can be cloned cheaply as they are smart pointers like `Rc`)
+ Use a `static` `sync::LazyLock` with either `Mutex` or [`Fragile`](https://crates.io/crates/fragile)

Both solutions are not ideal, but somehow you have to satisfy the constraints of the compiler.

**Reply by chrsoo** [reply]
> Use thread_local and then .with(|w| w.clone()) to access the widgets (since widgets can be cloned cheaply as they are smart pointers like Rc)

This... works! 🥇 Thanks a lot for all the help!

Not sure why this is not ideal though, given that that clones are cheap.

Speaking about ideal, I am more concerned with all the boxes. This is a screenshot from the GTK Debug tool, the last two boxes are from the two labels in the `view!` macro snippet above (I temporarily removed the entry)...
![image](https://github.com/user-attachments/assets/569e7a47-5e17-4c55-9418-37814e08e976)
![image](https://github.com/user-attachments/assets/908785ba-777e-42ae-9d02-30ca62ad2df4)

**Reply by AaronErhardt** [reply]
GTK often uses boxes internally, e.g. as a container to place child widgets inside. Yet, it seems to me like you have created most of the boxes yourself. In this case I think you should at least have created the following widgets yourself: Box -> Stack -> 3x Box -> Frame (probably one box as part of the frame) -> Box -> Stack -> 2x Box

In particular, I think you don't have to put the Frame into a Box and probably there are other places where you can just remove one Box and directly place the internal widget.

**Reply by chrsoo** [reply]
* The box wrapping the frame has been eliminated.
* I can't eliminate one of the two boxes after the first stack as it seems I cannot place the builder output in the root of the device property factory:
```rust
    view! {
        #[root]
        gtk::Box {
            set_orientation: gtk::Orientation::Vertical,
            // set_spacing: 10,
            self.props.widget() ->  // would be ideal if this could go in the root!
            &gtk::Box {
                set_orientation: gtk::Orientation::Vertical,
                set_spacing: 10,
            },
        },
        #[local_ref]
        returned_widget -> gtk::StackPage {
            set_name: self.device.name(),
            set_title: self.device.name(),
        }
    }
```
* Similar issues for the property item factory and here I also have the extra boxes introduced by the conditional widget:
```rust
    view! {
        #[root]
        gtk::Box {  // can't append the output of the conditional widget to the root
            set_orientation: gtk::Orientation::Horizontal,
            set_spacing: 10,
            append = match &self.item.value { 
                PropertyValue::Text(s) => {
                    gtk::Box {  // must return a single widget when using match
                        set_orientation: gtk::Orientation::Horizontal,
                        set_spacing: 10,
                        gtk::Box { //  `set_size_group` does not exist on gtk::Label
                            set_size_group: &PROP_COLUMN_1.with(|w| w.clone() ),
                            gtk::Label {
                                #[watch]
                                set_label: &self.item.name,
                            },
                        },
```
It would be great if it was possible to directly assign the match output to the root in the conditional widget, this would get rid of one more box.

**Reply by chrsoo** [reply]
Here is an annotated view of the UI:
![image](https://github.com/user-attachments/assets/f76a29e9-7b11-4725-94f2-ae43ecde509a)

**Reply by AaronErhardt** [reply]
Ok, the last to problems are actually limitations of the macro as far as I am aware. If you want you can open an issue for them, otherwise this will probably be forgotten.

**Reply by chrsoo** [reply]
I will open an issue tomorrow - thank you so much for all the help!

**Reply by chrsoo** [reply]
Created issue [690](https://github.com/Relm4/Relm4/issues/690) in the relm4 project.

**Reply by chrsoo** [reply]
... and I just discovered the "create issue from discussion button". Too late, next time... :-/
